experimental/dds/tree/docs/Future.md

```diff
@@ -0,0 +1,155 @@
+
+# Editing History
+
+The history is a sequence of Edits, but it is append only.
+Thus a logical edit of the history is actually done by creating a new Edit which modifies the current version of the document to be that which would have been produced by the alternative history.
+
+This functionality will be provided by a library (yet to be written) which uses SharedTree's history inspection and Editing APIs to create history modifying Edits, including metadata necessary to properly inspect and merge them with future history modifying Edits.
+This approach does not require extra functionality in SharedTree's core.
+This approach intentionally pushes the complex merge logic into code that is not part of the DDS to:
+
+-   reduce the code which can possibly impact consistency of documents between clients.
+-   reduce the computation needed to interpret the history (apply Edits): burden of hard merges goes on the client that requested the merge.
+-   minimize the feature set which must be supported long term for document compatibility.
+
+This design has some downsides as well:
+
+-   if used to make edits far back in the history, merge resolution can produce very bloated changes (large and expensive to store and apply)
+-   repeated history edits (like someone undoing and redoing a change a lot) can bloat history
+
+Because of these downsides, it may be desirable to limit use of history editing (it is intended to handle cases like undo and redo of recent edits),
+particularly when the initial edit requires Automatic Conflict Resolution, and/or is working on very large Edits or Edits which are not recent.
+
+These costs can be mitigated by encoding optimizations (Ex: allowing edits to refer to trees in snapshots and/or other edits to avoid duplication).
+
+If these mitigation are not enough, directly supporting history editing within SharedTree's Edits could be added without breaking existing documents or legacy support for the old approach
+(other than optionally reading the metadata when viewing history).
+This will be done if the future only if a cost/benefit analysis suggests it would be a good idea.
+
+## Example History Edit
+
+This example shows how to replace the middle of 3 edits.
+This generalizes to making arbitrary history edits by considering A, B, C, and X to be arbitrary (possibly empty) sequences of edits.
+
+We start with these changes:
+
+```mermaid
+graph LR
+A(A) --> B(B) --> C(C)
+```
+
+And want to end with changes:
+
+```mermaid
+graph LR
+A(A) --> X(X) --> C(C)
+```
+
+Since history is append only, we actually end with:
+
+```mermaid
+graph LR
+A(A) --> B(B) --> C(C) --> Merge(Merge)
+```
+
+### Low Level Commuting Merge
+
+If B commutes with C in the revision output from A, and X commutes with C in the revision output from A: the history edit is considered non-conflicted, since there is a clear way to construct the Merge edit.
```

Comment:
```suggestion
If B commutes with C in the revision output from A, and X commutes with C in the revision output from A, then the history edit is considered non-conflicted, since there is a clear way to construct the Merge edit.
```